### PR TITLE
RavenDB-16359 Add support for JsonPatch syntax for docs patching

### PR DESCRIPTION
**Pull request by lastav5**
https://issues.hibernatingrhinos.com/issue/RavenDB-16359

### CODE REVIEW

File: src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.AspNetCore.JsonPatch;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Session;
+using Raven.Client.Json.Serialization;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Client.Documents.Commands.Batches
+{
+    class JsonPatchCommandData : ICommandData
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs]:
```suggestion
    public class JsonPatchCommandData : ICommandData
```

File: src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        public string Id { get; }
+        public JsonPatchDocument JsonPatch { get; }
+        public string Name { get; } = null;
+        public string ChangeVector { get; }
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs:17]: Who is using these?

File: src/Raven.Client/Documents/Operations/JsonPatchOperation.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            public JsonPatchCommand(string id, JsonPatchDocument jsonPatchDocument)
+            {
+                _id = id;
+                _jsonPatchDocument = jsonPatchDocument;
+            }
+
+            public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+            {
+                url = $"{node.Url}/databases/{node.Database}/json-patch?id={_id}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/JsonPatchOperation.cs]: URL encode the id here.

File: src/Raven.Client/Documents/Operations/JsonPatchOperation.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+                            var serializer = DocumentConventions.Default.Serialization.CreateSerializer(new CreateSerializerOptions { TypeNameHandling = TypeNameHandling.None });
+                            foreach (var operation in _jsonPatchDocument.Operations)
+                            {
+                                writer.WriteValue(BlittableJsonToken.EmbeddedBlittable, DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(operation, ctx, serializer));
+                            }
+
+                            writer.WriteEndArray();
+                            writer.WriteEndObject();
+                        }
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/JsonPatchOperation.cs]: You are doing a lot of work in really awkward manner here, mixing multiple serialization styles.

File: src/Raven.Client/Documents/Session/Operations/BatchOperation.cs [hunk 5]
```diff
@@ -385,6 +388,58 @@ private void HandlePatch(BlittableJsonReaderObject batchResult)
             }
         }
 
+        private void HandleJsonPatch(BlittableJsonReaderObject batchResult)
+        {
+            if (batchResult.TryGet(nameof(PatchStatus), out string statusAsString) == false)
+                ThrowMissingField(CommandType.JsonPatch, nameof(PatchStatus));
+
+            if (Enum.TryParse(statusAsString, ignoreCase: true, out PatchStatus status) == false)
+                ThrowMissingField(CommandType.JsonPatch, nameof(PatchStatus));
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/Operations/BatchOperation.cs]: Not missing field, a wrong value, need better error here.

File: src/Raven.Server/Documents/Handlers/BatchRequestParser.cs [hunk 6]
```diff
@@ -884,6 +907,8 @@ private enum CommandPropertyName
             FromEtl
 
             // other properties are ignore (for legacy support)
+            ,
+            JsonPatch
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/BatchRequestParser.cs]: The comment is misleading, and the `,` is in the wrong place.

File: src/Raven.Server/Documents/Handlers/JsonPatchHandler.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+            using (ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+            {
+                var blittable = await context.ReadForMemoryAsync(RequestBodyStream(), "json-patch");
+                if (blittable == null)
+                    throw new BadRequestException("Missing JSON content.");
+
+                var commands = JsonPatchCommand.Parse(blittable);
+                var jsonPatchCommand = new JsonPatchCommand(id, commands, returnDocument: false, context);
+                await Database.TxMerger.Enqueue(jsonPatchCommand);
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/JsonPatchHandler.cs:29]: Aren't we returning anything to the user here?

File: src/Raven.Server/Documents/Patch/JsonPatchParser.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            
+        }
+
+        class JsonPatch
+        {
+            private string operation;
+            private string path;
+            private string value;
+        }
+    }
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/JsonPatchParser.cs]: What is this for?

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            _returnDocument = returnDocument;
+        }
+
+        protected override long ExecuteCmd(DocumentsOperationContext context)
+        {
+            var document = context.DocumentDatabase.DocumentsStorage.Get(context, _id);
+            
+            if (document == null)
+            {
+                //TODO: Return patchResult with PatchStatus.DocumentDoesNotExist instead of error?
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Should handle this, no?

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+                modifiedCollections?.Add(_patchResult.Collection);
+
+            var patchReply = new DynamicJsonValue
+            {
+                [nameof(BatchRequestParser.CommandData.Id)] = _id,
+                [nameof(BatchRequestParser.CommandData.ChangeVector)] = _patchResult.ChangeVector,
+                [nameof(Constants.Documents.Metadata.LastModified)] = _patchResult.LastModified,
+                [nameof(BatchRequestParser.CommandData.Type)] = nameof(CommandType.JsonPatch),
+                [nameof(PatchStatus)] = _patchResult.Status,
+                [nameof(PatchResult.Debug)] = _patchResult.Debug //TODO: needed?
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Probably not needed, no

Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: You are mixing the `nameof` fields here, which is confusing. What is this being read into, use that

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+                        result = obj;
+                        break;
+                    default:
+                        throw new ArgumentException($"Cannot reach target location. Failed to fetch '{name}' in path '{fullPath}' for operation '{Type}'");
+                }
+            }
+
+            public void Test(object objectAtPath, object value)
+            {
+                //No need to check type of objectAtPath because all JsonReader objects implement Equals
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Are you operating on `JsonReader` here, pretty sure that this isn't the case.
What if you are trying to compare `BlitableJsonObject` ?

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+                        arr.EnsureArrayModifiable();
+                        if (prop == "-")
+                        {
+                            arr.Modifications.Items.Add(value);
+                            return;
+                        }
+                        if (int.TryParse(prop, out var index) == false)
+                            throw new ArgumentException($"Expected an index but got a '{prop}' at path {OriginalPath}");
+                        
+                        //TODO: check and throw our own error for out of bounds argument?
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Still need to handle this

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs (hunk 10 above)
Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: I followed the same code as in `PatchDocumentCommand`

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs (hunk 9 above)
Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: JsonPatch's library throws an error when we run it with a `null` object.
The Raven Patch returns `PatchStatus.DocumentDoesNotExist`.
What should we do here?

File: src/Raven.Server/Documents/Patch/JsonPatchParser.cs (hunk 8 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Patch/JsonPatchParser.cs]: deleted

File: src/Raven.Server/Documents/Handlers/BatchRequestParser.cs (hunk 6 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/BatchRequestParser.cs]: done

File: src/Raven.Client/Documents/Operations/JsonPatchOperation.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Operations/JsonPatchOperation.cs]: Changed.

File: src/Raven.Client/Documents/Operations/JsonPatchOperation.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Operations/JsonPatchOperation.cs]: Done.

File: src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs:17]: ICommandData forces to implement them

File: src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Commands/Batches/JsonPatchCommandData.cs]: Done.

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs (hunk 12 above)
Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Done

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs (hunk 10 above)
Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Removed

Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Done

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs (hunk 12 above)
Comment (lastav5) [reply src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs]: Done

File: src/Raven.Server/Documents/Handlers/JsonPatchHandler.cs (hunk 7 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/JsonPatchHandler.cs:29]: Fixed

File: src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+                    if (member is BlittableJsonReaderObject nested)
+                        root = nested;
+                    else if (member is BlittableJsonReaderArray arr)
+                    {
+                        if(i == paths.Length - 2)
+                            return arr;
+                        root = arr;
+                    }
+                    else
+                        throw new ArgumentException($"Cannot reach target location. Failed to fetch '{paths[i+1]}' in path '{fullPath}' for operation '{Type}' because element at {paths[i]} is not an object or a collection");
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs:158]: You are not specifying what is the document id here.
Better to wrap the whole calling function with a `try/catch` that would add such details

**Review by ayende — COMMENTED** [review]
Aside from the error handling, looks great

**Review by ppekrol — APPROVED** [review]

**Pull request merged**